### DISCUSSION TITLE
CaseReducer and modularized path states

### DISCUSSION THREAD

**Discussion by seanmrich** [pointfreeco/swift-composable-architecture#2804] [Q&A]
I love the simplification of path reducers and associated `NavigationStack` destinations in v1.8. However, I've had some trouble integrating into an app. It relates to the issue of how a child feature can push a path element onto the stack. The child feature can't see the `Path` state if it's defined in a separate module without importing the parent which causes a circular reference. From previous discussions, there are two recommendations to handling this:
* Use `NavigationLink(state:label:)` in the child feature to push a new path element. To avoid circular references between the parent and child features, create a separate module that holds the `State` types for the path and all associated child features. 
* Use `Button` instead of `NavigationLink` and send a delegate action requesting a new element for the path. The parent feature monitors for those delegate actions and manages the `StackState` itself.

I've been using the first method, and it isn't clear how to use the new 1.8 features to clean up the path reducer. Here's the shape of the existing type where `RootPathState` is declared in a separate module `StatesModule`:
```swift
@Reducer
struct Path {
  typealias State = StatesModule.RootPathState
  enum Action {
    case child(Child.Action)
  }
  var body: some ReducerOf<Self> {
    Scope(state: \.child, action: \.child) {
      Child()
    }
  }
}
```

If I want to pare this down using the new 1.8 hotness:
```swift
@Reducer
enum Path {
  typealias State = StatesModule.RootPathState
  case child(Child)
}
```

Much better 👍 and this compiles, but using it in the parent reducer raises an error because `Path.State` didn't get conformance to `CaseReducerState`:
```
var body: some ReducerOf<Self> {
  Reduce { ... }
    .forEach(\.path, action: \.path)
     ^ Error: Instance method `forEach(_:action:)` requires that `Parent.Path.State` conform to `CaseReducerState`
}
```

It makes sense that the `@Reducer` macro doesn't fiddle with an explicitly declared `State` type, but otherwise using the `NavigationLink` method described above doesn't seem possible. When using this sugared version of path declaration, is it necessary to use `Button`s for pushing navigation elements?

**Comment by stephencelis** [comment]
@seanmrich Can you extend the `State` to conform in the module containing `Path`?

```swift
@Reducer
enum Path {
  typealias State = StatesModule.RootPathState
  case child(Child)
}

extension Path.State: CaseReducerState {
  typealias StateReducer = Path
}
```

(You may need to move the extension to a separate file due to some macro bugs in Swift 5.9/10.)

This should unlock the shorthand `forEach` syntax as well as `store.case`.

Alternately, you can do the longhand `forEach` but refer to `Path.body` as the child reducer:

```swift
.forEach(\.path, action: \.path) {
  Path.body
}
```

**Reply by seanmrich** [reply]
Thanks, @stephencelis. Your first solution works to clear the errors from the `forEach`. I wasn't sure about conforming to `CaseReducerState` explicitly given the notice in the doc comments for `CaseReducer` to not conform explicitly.

However, I am getting errors in the macro expansion for `Path`. In the `scope(_:)` function, it switches on `store.state`, creating a `CaseScope` with a store for the `child` case. At that scope operation, there are three errors:
* Cannot convert value of type 'KeyPath<Root.Path.State, (ChildState) -> RootPathState>' to expected argument type 'KeyPath<Root.Path.State, ((ChildState) -> RootPathState)?>'
* Cannot convert value of type 'Store<(ChildState) -> RootPathState, ChildFeature.Action>' to expected argument type 'StoreOf<ChildFeature>'
* Key path cannot refer to enum case 'child'

Here's the code:
<details>
  <summary>Feature States</summary>

```swift
@ObservableState
public struct ChildState {}

@ObservableState
@CasePathable
public enum RootPathState {
  case child(ChildState)
}
```

</details>

<details>
  <summary>Child</summary>

```swift
@Reducer
public struct ChildFeature {
  public typealias State = ChildState
  public init() {}
}

public struct ChildFeatureView: View {
  let store: StoreOf<ChildFeature>
  
  public init(store: StoreOf<ChildFeature>) {
    self.store = store
  }
  
  public var body: some View {
    EmptyView()
  }
}
```

</details>

<details>
  <summary>Root</summary>

```swift
@Reducer
public struct Root {
  @Reducer
  public enum Path {
    public typealias State = RootPathState
    case child(ChildFeature)
  }
  
  @ObservableState
  public struct State {
    public var path: StackState<Path.State>
  }
  public enum Action {
    case path(StackAction<Path.State, Path.Action>)
  }
  public var body: some ReducerOf<Self> {
    EmptyReducer()
      .forEach(\.path, action: \.path)
  }
}

public struct RootView: View {
  @Bindable var store: StoreOf<Root>
  
  public init(store: StoreOf<Root>) {
    self.store = store
  }
  
  public var body: some View {
    NavigationStack(path: $store.scope(state: \.path, action: \.path)) {
      EmptyView()
    } destination: { pathStore in
      switch pathStore.case {
      case .child(let childStore):
        ChildFeatureView(store: childStore)
      }
    }
  }
}
```

</details>

Your second solution didn't work for me. I get an error at the `forEach` action key path: `Cannot infer key path type from context; consider explicitly specifying a root type`. If I fully specify the key paths as `forEach(\State.path, \Action.path)` I get two errors:
* Key path cannot refer to enum case 'path'
* No exact matches in call to instance method 'forEach'

Also, when you use a full `Path` declaration like below, I keep getting an error at the `NavigationStack` declaration that the `State` could not be inferred. 
```swift
@Reducer
struct Path {
  @ObservableState
  enum State {
    case child(Child.State)
  }
  enum Action {
    case child(Child.Action)
  }
  var body: some ReducerOf<Self> {
    Scope(state: \.child, action: \.child) {
      Child()
    }
  }
}
```

A caveat to these errors: I tried dozens of different configurations in a test project to work this out. The compiler really struggles with the macro expansions. You'll think there's some type errors, but restarting Xcode and doing a clean build clears them out. The results above seem to repeatable with clean builds, though.

I'm using Xcode 15.3 beta 3 for these experiments.

**Reply by NikolayGenov** [reply]
Is there any resolution on that ? @seanmrich ?

**Reply by seanmrich** [reply]
I’m afraid not. The approach I described was just too flaky. I developed a button that visually simulates a navigation link and sends a delegate action. The root reducer plucks out those delegate messages and populates the navigation path. I’ve found it to be much more reliable and you can keep your state objects in their own module rather than creating a separate module that houses them all. The features remain well isolated.

**Reply by stephencelis** [reply]
While finicky, this approach worked for me:

```swift
@Reducer struct Feature1 {
  @ObservableState
  struct State: Equatable {}
  enum Action {}
  let body = EmptyReducer<State, Action>()
}
@Reducer struct Feature2 {
  struct State: Equatable {}
  enum Action {}
  let body = EmptyReducer<State, Action>()
}

@dynamicMemberLookup
@CasePathable
@ObservableState
enum PathState: Equatable {
  case feature1(Feature1.State)
  case feature2(Feature2.State)
}

@Reducer
enum Path {
  case feature1(Feature1)
  case feature2(Feature2)

  typealias State = PathState
}

extension PathState: CaseReducerState {
  typealias StateReducer = Path
}
```

**Reply by HerashchenkoElena** [reply]
@stephencelis I have tried to implement your approach and it’s working well with IOS 17, but unfortunately for IOS 16 there is no navigation to the child and I have an error
```
A NavigationLink is presenting a value of type “Component” but there is no matching navigationDestination declaration visible from the location of the link. The link cannot be activated.
Note: Links search for destinations in any surrounding NavigationStack, then within the same column of a NavigationSplitView.
```

Any suggestions how to solve it?

**Reply by stephencelis** [reply]
@HerashchenkoElena You'd need to provide a repro to debug further. It's not immediately clear to me why you would get such a warning in iOS 16 and not 17 with the same code.

**Reply by HerashchenkoElena** [reply]
@stephencelis Here is an example that represents this problem
[TCA.zip](https://github.com/pointfreeco/swift-composable-architecture/files/14601445/TCA.zip)

**Reply by stephencelis** [reply]
@HerashchenkoElena This looks like a behavior change in vanilla SwiftUI when it comes to its navigation tools. The issue appears to be in AppFeature.swift, where two `NavigationStack`s wrap the underlying feature views, but those feature views also use `NavigationStack`s. SwiftUI does not support multiple nested `NavigationStack`s in the same view hierarchy, and the behavior seems undefined across OS releases. To fix I think you can remove the outer `NavigationStack`s in AppFeature.swift and things should work!

**Reply by HerashchenkoElena** [reply]
@stephencelis Yes, thank you! It works!
